GuardianConnect/Categories/NSObject+Dictionary.m

```diff
@@ -0,0 +1,95 @@
+//
+//  NSObject+Dictionary.m
+//  GuardianConnect
+//
+//  Created by Kevin Bradley on 7/9/21.
+//  Copyright © 2021 Sudo Security Group Inc. All rights reserved.
+//
+
+#import <GuardianConnect/NSObject+Dictionary.h>
+#import <objc/runtime.h>
+
+@implementation NSDictionary (String)
+
+- (NSString *)JSONRepresentation {
+    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:self options:NSJSONWritingPrettyPrinted error:nil];
+    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+}
+
+@end
+
+@implementation NSObject (Dictionary)
+
+- (NSArray *)propertiesForClass:(Class)clazz {
+    u_int count;
+    objc_property_t* properties = class_copyPropertyList(clazz, &count);
+    NSMutableArray* propArray = [NSMutableArray arrayWithCapacity:count];
+    for (int i = 0; i < count ; i++)
+    {
+        const char* propertyName = property_getName(properties[i]);
+        NSString *propName = [NSString  stringWithCString:propertyName encoding:NSUTF8StringEncoding];
+        [propArray addObject:propName];
+    }
+    free(properties);
+    return propArray;
+}
+
+- (NSArray *)properties {
+    u_int count;
+    objc_property_t* properties = class_copyPropertyList(self.class, &count);
+    NSMutableArray* propArray = [NSMutableArray arrayWithCapacity:count];
+    for (int i = 0; i < count ; i++)
+    {
+        const char* propertyName = property_getName(properties[i]);
+        NSString *propName = [NSString  stringWithCString:propertyName encoding:NSUTF8StringEncoding];
+        [propArray addObject:propName];
+    }
+    free(properties);
+    return propArray;
+}
+
+- (id)valueForUndefinedKey:(NSString *)key {
+    GRDLog(@"in value for undefined key: %@", key);
+    return nil;
+}
+
+/*
+ 
+ This extremely useful function will take the current class (rather than NSObject), convert all of its properties
+ and convert them into NSDictionary representations that /should/ be JSON friendly. This is mainly used to change SKProducts
+ into a useful dictionary to feed to the API for partner product ID's.
+ 
+ */
+
+- (NSDictionary *)dictionaryRepresentation {
+    __block NSMutableDictionary *dict = [NSMutableDictionary new];
+    Class cls = NSClassFromString([self valueForKey:@"className"]); //this is how we hone in our the properties /just/ for our specific class rather than NSObject's properties.
```

Comment:
is `[self valueForKey:@"className"]` pointing to the correct thing here at runtime? Would it not point at the overall NSObject instance of the JSON blob?